main/channel.c

```diff
@@ -708,16 +708,75 @@ static void ast_channel_destructor(void *obj);
 static void ast_dummy_channel_destructor(void *obj);
 static int ast_channel_by_uniqueid_cb(void *obj, void *arg, void *data, int flags);
 
+static struct ast_channel *ast_channel_callback_internal(ao2_callback_data_fn *cb_fn, void *arg,
+		void *data, int ao2_flags)
+{
+	return ao2_callback_data(channels, ao2_flags, cb_fn, arg, data);
+}
+
+struct ast_channel *ast_channel_callback(ao2_callback_data_fn *cb_fn, void *arg,
+		void *data, int ao2_flags)
+{
+	ast_log(LOG_WARNING, "Using ast_channel_callback() directly may lead to deadlocks.  Please update your module.\n");
+
+	return ast_channel_callback_internal(cb_fn, arg, data, ao2_flags);
+}
+
+struct ast_channel *ast_channel_callback_safe(ao2_callback_data_fn *cb_fn, const char *search_type, const char *search_type_data, const char *search, void *data_additional, int ao2_flags)
+{
+	struct ast_channel *channel = NULL;
+	char *l_search = (char *) search;
+	struct ast_channel_callback_data callback_data = {
+		.data_additional	= data_additional,
+		.final_match_result	= 0,
+	};
+
+	if (ast_strlen_zero(search)) {
+		return NULL;
+	}
+
+        ast_debug(5, "Try to find channel by %s: %s\n", search_type, search_type_data);
+
+	for (;;) {
+		callback_data.final_match_result = 0;
+
+		/* It's REQUIRED that the passed in cb_fn does a trylock on the channel it wants to operate on, and *
+		 * then if the trylock fails: stop the search and set callback_data.final_match_result as necessary */
+		channel = ast_channel_callback_internal(cb_fn, l_search, &callback_data, ao2_flags);
+
+		/* In order to avoid deadlocks, ast_channel_callback must operate quickly and without holding *
+		 * locks for any length of time longer than necessary to return a channel pointer	      */
+
+		/* ast_channel_callback will lock the 'channels' (main channel list) *
+		 * mutex.  We don't want to hold onto the channels mutex AND block   *
+		 * on a channel lock, if we cannot grab a channel lock right away    */
+
+		if (callback_data.final_match_result == CMP_RETRY_NEEDED) {
+			/* We ran into a possible deadlock, try again */
+			ast_debug(3, "Avoid deadlock trying to find channel by %s: %s\n", search_type, search_type_data);
+			usleep(100);
+			continue;
+		}
+
+		/* We found a match or the search fully completed through all channels */
+		break;
+	}
+
+	return channel;
+}
+
 static int does_id_conflict(const char *uniqueid)
 {
 	struct ast_channel *conflict;
+        char *l_uniqueid = (char *) uniqueid;
```

Comment:
You're pointlessly casting to the same thing here

Reply:
It's dropping the const qualifier so that you don't have to cast it for the callback.